src/WebJobs.Extensions.DurableTask/Listener/TaskActorShim.cs

```diff
@@ -0,0 +1,323 @@
+
+        private Dictionary<Guid, Task> currentReentrantRequests = new Dictionary<Guid, Task>();
+
+        private int remainingReceiveLimit = 20; // TODO make configurable and/or compute based on state size
+
+        public TaskActorShim(DurableTaskExtension config, string schedulerId)
+            : base(config)
+        {
+            this.SchedulerId = schedulerId;
+            this.ActorId = GetActorIdFromSchedulerId(schedulerId);
+            this.context = new DurableActorContext(config, this.ActorId);
+        }
+
+        public override DurableCommonContext Context => context;
+
+        public string SchedulerId { get; private set; }
+
+        public ActorId ActorId { get; private set; }
+
+        public static string GetSchedulerIdFromActorId(ActorId actor)
+        {
+            return $"@{@actor.ActorClass}@{actor.ActorKey}";
+        }
+
+        public static ActorId GetActorIdFromSchedulerId(string schedulerId)
+        {
+            var pos = schedulerId.IndexOf('@', 1);
+            var actorClass = schedulerId.Substring(1, pos - 1);
+            var actorKey = schedulerId.Substring(pos + 1);
+            return new ActorId(actorClass, actorKey);
+        }
+
+        public override RegisteredFunctionInfo GetFunctionInfo()
+        {
+            FunctionName actorFunction = new FunctionName(this.Context.FunctionName);
+            return this.Config.GetActorInfo(actorFunction);
+        }
+
+        public override string GetStatus()
+        {
+            if (this.lastStartedOperation.IsCompleted)
+            {
+                return "Available";
+            }
+            else
+            {
+                if (this.context.State.Queue != null)
+                {
+                    return $"Busy ({this.context.State.Queue.Count} Queued)";
+                }
+                else
+                {
+                    return "Busy";
+                }
+            }
+        }
+
+        private void SignalShutdown()
+        {
+            MethodInfo trySetResult = typeof(TaskCompletionSource<object>).GetMethod("TrySetResult");
+            trySetResult.Invoke(this.shutDownSignal, new[] { (object)null });
+            this.shutDownSignal.TrySetResult(0);
+        }
+
+        public override void RaiseEvent(OrchestrationContext unused, string eventName, string serializedEventData)
+        {
+            this.Config.TraceHelper.ExternalEventRaised(
+                this.Context.HubName,
+                this.Context.Name,
+                this.Context.InstanceId,
+                eventName,
+                this.Config.GetIntputOutputTrace(serializedEventData),
+                this.Context.IsReplaying);
+
+            if (eventName == "op")
+            {
+                var operationMessage = JsonConvert.DeserializeObject<OperationMessage>(serializedEventData);
+
+                if (this.lastStartedOperation.IsCompleted && this.remainingReceiveLimit > 0)
```

Comment:
yes, there is an interesting design question here. I will add comments. 

The point is that we need to prevent the history from growing too large. I currently do that by counting how many requests were received, and calling continueAsNew after handling 20 requests. But we probably want to take another look at this heuristic once we get some perf experiments going.

Reply:
I think I will rename this to `numberOperationsProcessed` which should be more readable.